core/src/main/scala/shapeless/package.scala

```diff
@@ -59,6 +59,60 @@ package object shapeless {
   type ∃[P[_]] = P[T] forSome { type T }
   type ∀[P[_]] = ¬[∃[({ type λ[X] = ¬[P[X]]})#λ]]
 
+  //Type logic-gates
+  trait ||[A, B] extends Serializable {
+    def a: Option[A]
+    def b: Option[B]
+  }
+
```

Comment:
The original motivation comes from line 252 of this [here](https://github.com/to-ithaca/iliad/blob/master/core/src/main/scala/iliad/algebra/matrix.scala)

I essentially need evidence that either `M <: Nat`, `N <: Nat` or both are greater than a certain number as I don't want to be able to call the method unless that is the case. 

I'm not sure these belong in shapeless which is why this is a speculative PR. I'm entirely happy with leaving it in our own codebase.